# Intro to MultiIndex Module
- Multiple layers, Multiple levels, Multiple tiers

In [1]:
import pandas as pd

Interestingly, prices of Bigmac kinda represents country's economy indicator.

In [3]:
bigmac = pd.read_csv('Data/bigmac.csv')
bigmac.head()

Date    Country  Price in US Dollars
0  1/2016  Argentina                 2.39
1  1/2016  Australia                 3.74
2  1/2016     Brazil                 3.35
3  1/2016    Britain                 4.22
4  1/2016     Canada                 4.14

In [4]:
bigmac.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 652 entries, 0 to 651
Data columns (total 3 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Date                 652 non-null    object 
 1   Country              652 non-null    object 
 2   Price in US Dollars  652 non-null    float64
dtypes: float64(1), object(2)
memory usage: 15.4+ KB


In [6]:
bigmac.isnull().sum()

Date                   0
Country                0
Price in US Dollars    0
dtype: int64

## Optimization Points
+ Date are in object datatype. If necessary, we need to change to Datetime.
+ Prices are in object (String), so we need to change them to float to allow us make calculations.

In [8]:
bigmac = pd.read_csv('Data/bigmac.csv', parse_dates=['Date'])
bigmac.head()

Date    Country  Price in US Dollars
0 2016-01-01  Argentina                 2.39
1 2016-01-01  Australia                 3.74
2 2016-01-01     Brazil                 3.35
3 2016-01-01    Britain                 4.22
4 2016-01-01     Canada                 4.14

------

# 2) Create a MultiIndex with the `set_index` Method

In [31]:
bigmac = pd.read_csv('Data/bigmac.csv', parse_dates=['Date'])
bigmac.head(3)

Date    Country  Price in US Dollars
0 2016-01-01  Argentina                 2.39
1 2016-01-01  Australia                 3.74
2 2016-01-01     Brazil                 3.35

### We can have multiIndex with 2 layers, 3 layers, many more, etc

In [27]:
bigmac.set_index(keys=['Date', 'Country'])
bigmac.set_index(keys=['Country', 'Date'])

Price in US Dollars
Country       Date                           
Argentina     2016-01-01                 2.39
Australia     2016-01-01                 3.74
Brazil        2016-01-01                 3.35
Britain       2016-01-01                 4.22
Canada        2016-01-01                 4.14
...                                       ...
Turkey        2010-01-01                 3.83
UAE           2010-01-01                 2.99
Ukraine       2010-01-01                 1.83
United States 2010-01-01                 3.58
Uruguay       2010-01-01                 3.32

[652 rows x 1 columns]

### TIPS: Column with less values should be the most Outer Layer 
In our case => Date

In [28]:
bigmac.nunique()

Date                    12
Country                 58
Price in US Dollars    330
dtype: int64

In [32]:
bigmac = bigmac.set_index(keys=['Date', 'Country'])
bigmac.head(3)

Price in US Dollars
Date       Country                       
2016-01-01 Argentina                 2.39
           Australia                 3.74
           Brazil                    3.35

## `sort_index()` after multi indexing

In [36]:
bigmac = bigmac.sort_index() # sort by date, then by country 
bigmac.head(3)

Price in US Dollars
Date       Country                       
2010-01-01 Argentina                 1.84
           Australia                 3.98
           Brazil                    4.76

### Index and Names

In [38]:
bigmac.index

bigmac.index.names

FrozenList(['Date', 'Country'])

In [39]:
type(bigmac.index)

pandas.core.indexes.multi.MultiIndex

### calling `.index` returns tuple for multi Index

In [40]:
bigmac.index

MultiIndex([('2010-01-01',      'Argentina'),
            ('2010-01-01',      'Australia'),
            ('2010-01-01',         'Brazil'),
            ('2010-01-01',        'Britain'),
            ('2010-01-01',         'Canada'),
            ('2010-01-01',          'Chile'),
            ('2010-01-01',          'China'),
            ('2010-01-01',       'Colombia'),
            ('2010-01-01',     'Costa Rica'),
            ('2010-01-01', 'Czech Republic'),
            ...
            ('2016-01-01',    'Switzerland'),
            ('2016-01-01',         'Taiwan'),
            ('2016-01-01',       'Thailand'),
            ('2016-01-01',         'Turkey'),
            ('2016-01-01',            'UAE'),
            ('2016-01-01',        'Ukraine'),
            ('2016-01-01',  'United States'),
            ('2016-01-01',        'Uruguay'),
            ('2016-01-01',      'Venezuela'),
            ('2016-01-01',        'Vietnam')],
           names=['Date', 'Country'], length=652)

In [41]:
bigmac.index[0]

(Timestamp('2010-01-01 00:00:00'), 'Argentina')

In [44]:
date, name = bigmac.index[0]
print('date: ', date)
print('Name: ', name)

date:  2010-01-01 00:00:00
Name:  Argentina


------

# 3) Extract Index Level Values with the `get_level_values` Method

In [48]:
bigmac = pd.read_csv('Data/bigmac.csv', parse_dates=['Date'], index_col=['Date', 'Country'])
bigmac = bigmac.sort_index()
bigmac.head(3)

Price in US Dollars
Date       Country                       
2010-01-01 Argentina                 1.84
           Australia                 3.98
           Brazil                    4.76

In [55]:
bigmac.index.get_level_values('Date')
bigmac.index.get_level_values(0)

DatetimeIndex(['2010-01-01', '2010-01-01', '2010-01-01', '2010-01-01',
               '2010-01-01', '2010-01-01', '2010-01-01', '2010-01-01',
               '2010-01-01', '2010-01-01',
               ...
               '2016-01-01', '2016-01-01', '2016-01-01', '2016-01-01',
               '2016-01-01', '2016-01-01', '2016-01-01', '2016-01-01',
               '2016-01-01', '2016-01-01'],
              dtype='datetime64[ns]', name='Date', length=652, freq=None)

In [57]:
bigmac.index.get_level_values(1)
bigmac.index.get_level_values('Country')

Index(['Argentina', 'Australia', 'Brazil', 'Britain', 'Canada', 'Chile',
       'China', 'Colombia', 'Costa Rica', 'Czech Republic',
       ...
       'Switzerland', 'Taiwan', 'Thailand', 'Turkey', 'UAE', 'Ukraine',
       'United States', 'Uruguay', 'Venezuela', 'Vietnam'],
      dtype='object', name='Country', length=652)

------

# 4) Change Index Level Name with the `set_names` Method

In [84]:
bigmac = pd.read_csv('Data/bigmac.csv', parse_dates=['Date'], index_col=['Date', 'Country'])
bigmac = bigmac.sort_index()
bigmac.head()

Price in US Dollars
Date       Country                       
2010-01-01 Argentina                 1.84
           Australia                 3.98
           Brazil                    4.76
           Britain                   3.67
           Canada                    3.97

In [85]:
bigmac.index

MultiIndex([('2010-01-01',      'Argentina'),
            ('2010-01-01',      'Australia'),
            ('2010-01-01',         'Brazil'),
            ('2010-01-01',        'Britain'),
            ('2010-01-01',         'Canada'),
            ('2010-01-01',          'Chile'),
            ('2010-01-01',          'China'),
            ('2010-01-01',       'Colombia'),
            ('2010-01-01',     'Costa Rica'),
            ('2010-01-01', 'Czech Republic'),
            ...
            ('2016-01-01',    'Switzerland'),
            ('2016-01-01',         'Taiwan'),
            ('2016-01-01',       'Thailand'),
            ('2016-01-01',         'Turkey'),
            ('2016-01-01',            'UAE'),
            ('2016-01-01',        'Ukraine'),
            ('2016-01-01',  'United States'),
            ('2016-01-01',        'Uruguay'),
            ('2016-01-01',      'Venezuela'),
            ('2016-01-01',        'Vietnam')],
           names=['Date', 'Country'], length=652)

### Change the names of Our Index
- need to be in the same order when passing `names = []` parameter

In [86]:
bigmac.index.set_names(names = ['Day', 'Location'])

MultiIndex([('2010-01-01',      'Argentina'),
            ('2010-01-01',      'Australia'),
            ('2010-01-01',         'Brazil'),
            ('2010-01-01',        'Britain'),
            ('2010-01-01',         'Canada'),
            ('2010-01-01',          'Chile'),
            ('2010-01-01',          'China'),
            ('2010-01-01',       'Colombia'),
            ('2010-01-01',     'Costa Rica'),
            ('2010-01-01', 'Czech Republic'),
            ...
            ('2016-01-01',    'Switzerland'),
            ('2016-01-01',         'Taiwan'),
            ('2016-01-01',       'Thailand'),
            ('2016-01-01',         'Turkey'),
            ('2016-01-01',            'UAE'),
            ('2016-01-01',        'Ukraine'),
            ('2016-01-01',  'United States'),
            ('2016-01-01',        'Uruguay'),
            ('2016-01-01',      'Venezuela'),
            ('2016-01-01',        'Vietnam')],
           names=['Day', 'Location'], length=652)

In [88]:
bigmac.index.set_names(names = ['Day', 'Location'], inplace=True)

In [90]:
bigmac.head(3)

Price in US Dollars
Day        Location                      
2010-01-01 Argentina                 1.84
           Australia                 3.98
           Brazil                    4.76

## If we want to change only one index of multiindex, we can pass using `level` parameter
+ `level = level_number`
+ or `level = 'current index name'`

In [99]:
 # only changing from Day to Date
bigmac.index.set_names(names='Date', level=0)
bigmac.index.set_names(names='Date', level='Day')

MultiIndex([('2010-01-01',      'Argentina'),
            ('2010-01-01',      'Australia'),
            ('2010-01-01',         'Brazil'),
            ('2010-01-01',        'Britain'),
            ('2010-01-01',         'Canada'),
            ('2010-01-01',          'Chile'),
            ('2010-01-01',          'China'),
            ('2010-01-01',       'Colombia'),
            ('2010-01-01',     'Costa Rica'),
            ('2010-01-01', 'Czech Republic'),
            ...
            ('2016-01-01',    'Switzerland'),
            ('2016-01-01',         'Taiwan'),
            ('2016-01-01',       'Thailand'),
            ('2016-01-01',         'Turkey'),
            ('2016-01-01',            'UAE'),
            ('2016-01-01',        'Ukraine'),
            ('2016-01-01',  'United States'),
            ('2016-01-01',        'Uruguay'),
            ('2016-01-01',      'Venezuela'),
            ('2016-01-01',        'Vietnam')],
           names=['Date', 'Location'], length=652)

In [98]:
# only changing from Location to Country
bigmac.index.set_names(names='Country', level=1) 
bigmac.index.set_names(names='Country', level='Location')

MultiIndex([('2010-01-01',      'Argentina'),
            ('2010-01-01',      'Australia'),
            ('2010-01-01',         'Brazil'),
            ('2010-01-01',        'Britain'),
            ('2010-01-01',         'Canada'),
            ('2010-01-01',          'Chile'),
            ('2010-01-01',          'China'),
            ('2010-01-01',       'Colombia'),
            ('2010-01-01',     'Costa Rica'),
            ('2010-01-01', 'Czech Republic'),
            ...
            ('2016-01-01',    'Switzerland'),
            ('2016-01-01',         'Taiwan'),
            ('2016-01-01',       'Thailand'),
            ('2016-01-01',         'Turkey'),
            ('2016-01-01',            'UAE'),
            ('2016-01-01',        'Ukraine'),
            ('2016-01-01',  'United States'),
            ('2016-01-01',        'Uruguay'),
            ('2016-01-01',      'Venezuela'),
            ('2016-01-01',        'Vietnam')],
           names=['Day', 'Country'], length=652)

In [100]:
bigmac.index.set_names(names=['Date', 'Country'], inplace=True)
bigmac.head(3)

Price in US Dollars
Date       Country                       
2010-01-01 Argentina                 1.84
           Australia                 3.98
           Brazil                    4.76

-------

# 5) The `sort_index` Method on a `MultiIndex DataFrame`

In [103]:
bigmac = pd.read_csv('Data/bigmac.csv', parse_dates=['Date'], index_col=['Date', 'Country'])
bigmac.head(3)

Price in US Dollars
Date       Country                       
2016-01-01 Argentina                 2.39
           Australia                 3.74
           Brazil                    3.35

In [109]:
bigmac.sort_index() # sort_index works in the sequence of index (Date => Country)
bigmac.sort_index(ascending=True)

bigmac.sort_index(ascending=False)  # all levels are sorting descending (Date => Country)

Price in US Dollars
Date       Country                           
2016-01-01 Vietnam                       2.67
           Venezuela                     0.66
           Uruguay                       3.74
           United States                 4.93
           Ukraine                       1.54
...                                       ...
2010-01-01 Canada                        3.97
           Britain                       3.67
           Brazil                        4.76
           Australia                     3.98
           Argentina                     1.84

[652 rows x 1 columns]

## What if we want to customize Sorting order for index?
+ we can pass boolean as a list

In [110]:
bigmac.sort_index(ascending=[True, False])

# ascending for Date, descending for Country

Price in US Dollars
Date       Country                           
2010-01-01 Uruguay                       3.32
           United States                 3.58
           Ukraine                       1.83
           UAE                           2.99
           Turkey                        3.83
...                                       ...
2016-01-01 Brazil                        3.35
           Belgium                       4.25
           Austria                       3.76
           Australia                     3.74
           Argentina                     2.39

[652 rows x 1 columns]

In [111]:
bigmac.sort_index(ascending=[False, True])

Price in US Dollars
Date       Country                           
2016-01-01 Argentina                     2.39
           Australia                     3.74
           Austria                       3.76
           Belgium                       4.25
           Brazil                        3.35
...                                       ...
2010-01-01 Turkey                        3.83
           UAE                           2.99
           Ukraine                       1.83
           United States                 3.58
           Uruguay                       3.32

[652 rows x 1 columns]

In [115]:
bigmac.sort_index()

bigmac.sort_index(ascending=True)
bigmac.sort_index(ascending=False)

bigmac.sort_index(ascending=[True, False])
bigmac.sort_index(ascending=[False, True], inplace=True)

In [116]:
bigmac.head(3)

Price in US Dollars
Date       Country                       
2016-01-01 Argentina                 2.39
           Australia                 3.74
           Austria                   3.76

## Sorting by Single Level using `level` parameter

In [126]:
bigmac.sort_index(level=0)
bigmac.sort_index(level='Date')

Price in US Dollars
Date       Country                           
2010-01-01 Argentina                     1.84
           Australia                     3.98
           Brazil                        4.76
           Britain                       3.67
           Canada                        3.97
...                                       ...
2016-01-01 Ukraine                       1.54
           United States                 4.93
           Uruguay                       3.74
           Venezuela                     0.66
           Vietnam                       2.67

[652 rows x 1 columns]

In [127]:
# when we sort on level of 1 or Country level, Date values are brought along (not necessarily sorted though)
bigmac.sort_index(level=1, ascending=False)
bigmac.sort_index(level='Country', ascending=False)

Price in US Dollars
Date       Country                       
2016-01-01 Vietnam                   2.67
2015-07-01 Vietnam                   2.75
2015-01-01 Vietnam                   2.81
2014-07-01 Vietnam                   2.83
2014-01-01 Vietnam                   2.84
...                                   ...
2012-07-01 Argentina                 4.16
2012-01-01 Argentina                 4.64
2011-07-01 Argentina                 4.84
2010-07-01 Argentina                 3.56
2010-01-01 Argentina                 1.84

[652 rows x 1 columns]

-----

# 6) Extract Rows from a `MultiIndex DataFrame` using `.loc` and `.iloc` Accessors

In [128]:
bigmac = pd.read_csv('Data/bigmac.csv', index_col=['Date', 'Country'], parse_dates=['Date'])
bigmac = bigmac.sort_index()
bigmac.head(3)

Price in US Dollars
Date       Country                       
2010-01-01 Argentina                 1.84
           Australia                 3.98
           Brazil                    4.76

The following syntax is a bit ambiguous because 2nd value can be either
+ Index name
+ or Column name.
This will lead to confusion. That's why we shouldn't use like that.

In [133]:
bigmac.loc['2010-01-01']

bigmac.loc['2010-01-01', 'Brazil']
bigmac.loc['2010-01-01', 'Price in US Dollars']

Date        Country       
2010-01-01  Argentina         1.84
            Australia         3.98
            Brazil            4.76
            Britain           3.67
            Canada            3.97
            Chile             3.18
            China             1.83
            Colombia          3.91
            Costa Rica        3.52
            Czech Republic    3.71
            Denmark           5.99
            Egypt             2.38
            Euro area         4.84
            Hong Kong         1.91
            Hungary           3.86
            Indonesia         2.24
            Israel            3.99
            Japan             3.50
            Latvia            3.09
            Lithuania         2.87
            Malaysia          2.08
            Mexico            2.50
            New Zealand       3.61
            Norway            7.02
            Pakistan          2.42
            Peru              2.81
            Philippines       2.21
            Poland          

## Correct Approach (Panda Convention) for `.loc` Accessor for MultiIndex
+ **1st Argument as tuple for ROW Index** (it can include index or multi indexes)
+ **2nd Arugment as tuple for COLUMN names**
+ **NOTE: if there is one value to pass in 1st Argument, make sure to use comma , to indicate that it is tuple.**

This is consistent and less ambiguous.

In [139]:
bigmac.loc[('2010-01-01', 'Brazil')]
bigmac.loc[('2010-01-01', 'Brazil'), 'Price in US Dollars']
bigmac.loc[('2010-01-01', 'Brazil'), ['Price in US Dollars', 'Price in US Dollars']]

# NOTE: if there is one value to pass in 1st Argument, make sure to use comma , to indicate that it is tuple.
bigmac.loc[('2010-01-01', )]

Price in US Dollars
Country                            
Argentina                      1.84
Australia                      3.98
Brazil                         4.76
Britain                        3.67
Canada                         3.97
Chile                          3.18
China                          1.83
Colombia                       3.91
Costa Rica                     3.52
Czech Republic                 3.71
Denmark                        5.99
Egypt                          2.38
Euro area                      4.84
Hong Kong                      1.91
Hungary                        3.86
Indonesia                      2.24
Israel                         3.99
Japan                          3.50
Latvia                         3.09
Lithuania                      2.87
Malaysia                       2.08
Mexico                         2.50
New Zealand                    3.61
Norway                         7.02
Pakistan                       2.42
Peru                           2.81
Philippines                    2.21
Poland                         2.86
Russia                         2.34
Saudi Arabia                   2.67
Singapore                      3.19
South Africa                   2.46
South Korea                    2.98
Sri Lanka                      1.83
Sweden                         5.51
Switzerland                    6.30
Taiwan                         2.36
Thailand                       2.11
Turkey                         3.83
UAE                            2.99
Ukraine                        1.83
United States                  3.58
Uruguay                        3.32

In [140]:
bigmac.head(2)

Price in US Dollars
Date       Country                       
2010-01-01 Argentina                 1.84
           Australia                 3.98

## using `.iloc[]` for multiindex

In [149]:
bigmac.iloc[0]
bigmac.iloc[10]
bigmac.iloc[[10, 20, 30]]
bigmac.iloc[[10, 20, 30, 250]]|

Price in US Dollars
Date       Country                       
2010-01-01 Denmark                   5.99
           Malaysia                  2.08
           Singapore                 3.19
2012-07-01 Turkey                    4.52

-------

# 7) The `transpose` Method
- swap the indexes of the row labels and column labels

In [150]:
bigmac = pd.read_csv('Data/bigmac.csv', index_col=['Date', 'Country'], parse_dates=['Date'])
bigmac = bigmac.sort_index()
bigmac.head(3)

Price in US Dollars
Date       Country                       
2010-01-01 Argentina                 1.84
           Australia                 3.98
           Brazil                    4.76

In [152]:
bigmac = bigmac.transpose()
bigmac.head(3)

Date                2010-01-01                                              \
Country              Argentina Australia Brazil Britain Canada Chile China   
Price in US Dollars       1.84      3.98   4.76    3.67   3.97  3.18  1.83   

Date                                                    ...  2016-01-01  \
Country             Colombia Costa Rica Czech Republic  ... Switzerland   
Price in US Dollars     3.91       3.52           3.71  ...        6.44   

Date                                                                    \
Country             Taiwan Thailand Turkey   UAE Ukraine United States   
Price in US Dollars   2.08     3.09   3.41  3.54    1.54          4.93   

Date                                           
Country             Uruguay Venezuela Vietnam  
Price in US Dollars    3.74      0.66    2.67  

[1 rows x 652 columns]

## using the format of Best Practice tuple() for both row index and column labels

NOTE: Always remember:
+ **1st Argument as tuple for ROW Index** (it can include index or multi indexes)
+ **2nd Arugment as tuple for COLUMN names** (it can include index or multi indexes)

In [163]:

bigmac.loc[('Price in US Dollars',)]

bigmac.loc[('Price in US Dollars', ), ('2010-01-01', )]

bigmac.loc[('Price in US Dollars', ), ('2010-01-01', 'Canada')]

bigmac.loc[('Price in US Dollars', ), ('2010-01-01', 'Canada'): ('2010-01-01', 'Taiwan')] 
# to extract values, we need to tell pandas both multi indexes (Date and Country)

Date                2010-01-01                                                 \
Country                 Canada Chile China Colombia Costa Rica Czech Republic   
Price in US Dollars       3.97  3.18  1.83     3.91       3.52           3.71   

Date                                                   ...                \
Country             Denmark Egypt Euro area Hong Kong  ... Poland Russia   
Price in US Dollars    5.99  2.38      4.84      1.91  ...   2.86   2.34   

Date                                                                           \
Country             Saudi Arabia Singapore South Africa South Korea Sri Lanka   
Price in US Dollars         2.67      3.19         2.46        2.98      1.83   

Date                                           
Country             Sweden Switzerland Taiwan  
Price in US Dollars   5.51         6.3   2.36  

[1 rows x 33 columns]

----

# 8) The `swaplevel()` Method

In [165]:
bigmac = pd.read_csv('Data/bigmac.csv', index_col=['Date', 'Country'], parse_dates=['Date'])
bigmac = bigmac.sort_index()
bigmac.head(3)

Price in US Dollars
Date       Country                       
2010-01-01 Argentina                 1.84
           Australia                 3.98
           Brazil                    4.76

In our case, there are only two level of indexes. So pandas know to swap them accordingly.

**If we have more than two level of indexes, we need to tell pandas which one (`i` )to swap with which (`j`).**

In [176]:
bigmac.swaplevel()

bigmac.swaplevel('Date', 'Country')
bigmac.swaplevel('Country', 'Date')

bigmac.swaplevel(i=1, j=0)
bigmac.swaplevel(1,0)

Price in US Dollars
Country       Date                           
Argentina     2010-01-01                 1.84
Australia     2010-01-01                 3.98
Brazil        2010-01-01                 4.76
Britain       2010-01-01                 3.67
Canada        2010-01-01                 3.97
...                                       ...
Ukraine       2016-01-01                 1.54
United States 2016-01-01                 4.93
Uruguay       2016-01-01                 3.74
Venezuela     2016-01-01                 0.66
Vietnam       2016-01-01                 2.67

[652 rows x 1 columns]

In [178]:
bigmac = bigmac.swaplevel(1,0)
bigmac.head()

Price in US Dollars
Country   Date                           
Argentina 2010-01-01                 1.84
Australia 2010-01-01                 3.98
Brazil    2010-01-01                 4.76
Britain   2010-01-01                 3.67
Canada    2010-01-01                 3.97

------

# 9) The `.stack()` Method
**it moves Column based indexes to Row based indexes.**

In [231]:
world = pd.read_csv('Data/worldstats.csv', index_col=['country', 'year'])
world.head(3)

Population           GDP
country    year                           
Arab World 2015  392022276.0  2.530102e+12
           2014  384222592.0  2.873600e+12
           2013  376504253.0  2.846994e+12

In [183]:
world.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 11211 entries, ('Arab World', 2015) to ('Zimbabwe', 1960)
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Population  11211 non-null  float64
 1   GDP         11211 non-null  float64
dtypes: float64(2)
memory usage: 223.0+ KB


**After stacking, Population and GDP column are moved to Row based indexes.**

In [184]:
world.stack()

country     year            
Arab World  2015  Population    3.920223e+08
                  GDP           2.530102e+12
            2014  Population    3.842226e+08
                  GDP           2.873600e+12
            2013  Population    3.765043e+08
                                    ...     
Zimbabwe    1962  GDP           1.117602e+09
            1961  Population    3.876638e+06
                  GDP           1.096647e+09
            1960  Population    3.752390e+06
                  GDP           1.052990e+09
Length: 22422, dtype: float64

In [185]:
type(world.stack())

pandas.core.series.Series

### We can convert series to Dataframe using `.to_frame()`

In [186]:
world.stack().to_frame()

0
country    year                         
Arab World 2015 Population  3.920223e+08
                GDP         2.530102e+12
           2014 Population  3.842226e+08
                GDP         2.873600e+12
           2013 Population  3.765043e+08
...                                  ...
Zimbabwe   1962 GDP         1.117602e+09
           1961 Population  3.876638e+06
                GDP         1.096647e+09
           1960 Population  3.752390e+06
                GDP         1.052990e+09

[22422 rows x 1 columns]

-------

# 10) The `.unstack()` Method - Part 1
**move Row Based Indexes to Column Based Indexes**

In [187]:
world = pd.read_csv('Data/worldstats.csv', index_col=['country', 'year'])
world.head(3)

Population           GDP
country    year                           
Arab World 2015  392022276.0  2.530102e+12
           2014  384222592.0  2.873600e+12
           2013  376504253.0  2.846994e+12

In [190]:
s = world.stack()
s.head(3)

country     year            
Arab World  2015  Population    3.920223e+08
                  GDP           2.530102e+12
            2014  Population    3.842226e+08
dtype: float64

In [194]:
s.unstack()
s.unstack().unstack()
s.unstack().unstack().unstack().to_frame()

0
           year country                         
Population 1960 Afghanistan         8.994793e+06
                Albania                      NaN
                Algeria             1.112489e+07
                Andorra                      NaN
                Angola                       NaN
...                                          ...
GDP        2015 West Bank and Gaza  1.267740e+10
                World               7.343364e+13
                Yemen, Rep.                  NaN
                Zambia              2.120156e+10
                Zimbabwe            1.389294e+10

[28224 rows x 1 columns]

----

# 11) The `.unstack()` Method - Part 2

In [195]:
world = pd.read_csv('Data/worldstats.csv', index_col=['country', 'year'])
world.head(3)

Population           GDP
country    year                           
Arab World 2015  392022276.0  2.530102e+12
           2014  384222592.0  2.873600e+12
           2013  376504253.0  2.846994e+12

In [196]:
s = world.stack()
s.head(3)

country     year            
Arab World  2015  Population    3.920223e+08
                  GDP           2.530102e+12
            2014  Population    3.842226e+08
dtype: float64

In [198]:
s.to_frame().head(3)

0
country    year                         
Arab World 2015 Population  3.920223e+08
                GDP         2.530102e+12
           2014 Population  3.842226e+08

## We can unstack on different level based on requirements using `level` parameter

In [213]:
s.unstack(level=0)
s.unstack(0) # country

s.unstack(1) # year
s.unstack(2) # Popultion and GDP

# we can also use negative indexes (moving backwards)
s.unstack(-1) #Population and GDP
s.unstack(-2) # Year
s.unstack(-3) # country

country           Afghanistan       Albania       Algeria       Andorra  \
year                                                                      
1960 Population  8.994793e+06           NaN  1.112489e+07           NaN   
     GDP         5.377778e+08           NaN  2.723638e+09           NaN   
1961 Population  9.164945e+06           NaN  1.140486e+07           NaN   
     GDP         5.488889e+08           NaN  2.434767e+09           NaN   
1962 Population  9.343772e+06           NaN  1.169015e+07           NaN   
...                       ...           ...           ...           ...   
2013 GDP         2.004633e+10  1.278103e+10  2.097035e+11  3.249101e+09   
2014 Population  3.162751e+07  2.893654e+06  3.893433e+07           NaN   
     GDP         2.005019e+10  1.327796e+10  2.135185e+11           NaN   
2015 Population  3.252656e+07  2.889167e+06  3.966652e+07           NaN   
     GDP         1.919944e+10  1.145560e+10  1.668386e+11           NaN   

country                Angola  Antigua and Barbuda    Arab World  \
year                                                               
1960 Population           NaN                  NaN           NaN   
     GDP                  NaN                  NaN           NaN   
1961 Population           NaN                  NaN           NaN   
     GDP                  NaN                  NaN           NaN   
1962 Population           NaN                  NaN           NaN   
...                       ...                  ...           ...   
2013 GDP         1.249121e+11         1.200588e+09  2.846994e+12   
2014 Population  2.422752e+07         9.090000e+04  3.842226e+08   
     GDP         1.267751e+11         1.220976e+09  2.873600e+12   
2015 Population  2.502197e+07         9.181800e+04  3.920223e+08   
     GDP         1.026431e+11         1.297285e+09  2.530102e+12   

country             Argentina       Armenia  Aruba  ...    Uzbekistan  \
year                                                ...                 
1960 Population           NaN           NaN    NaN  ...           NaN   
     GDP                  NaN           NaN    NaN  ...           NaN   
1961 Population           NaN           NaN    NaN  ...           NaN   
     GDP                  NaN           NaN    NaN  ...           NaN   
1962 Population  2.128768e+07           NaN    NaN  ...           NaN   
...                       ...           ...    ...  ...           ...   
2013 GDP         6.239320e+11  1.112147e+10    NaN  ...  5.679566e+10   
2014 Population  4.298003e+07  3.006154e+06    NaN  ...  3.075770e+07   
     GDP         5.480549e+11  1.164444e+10    NaN  ...  6.313285e+10   
2015 Population           NaN  3.017712e+06    NaN  ...  3.129950e+07   
     GDP                  NaN  1.056140e+10    NaN  ...  6.673280e+10   

country               Vanuatu  Venezuela, RB       Vietnam  \
year                                                         
1960 Population           NaN   8.146845e+06           NaN   
     GDP                  NaN   8.607600e+09           NaN   
1961 Population           NaN   8.461684e+06           NaN   
     GDP                  NaN   8.923367e+09           NaN   
1962 Population           NaN   8.790590e+06           NaN   
...                       ...            ...           ...   
2013 GDP         8.017876e+08   3.713366e+11  1.712220e+11   
2014 Population  2.588830e+05            NaN  9.072890e+07   
     GDP         8.149546e+08            NaN  1.862047e+11   
2015 Population           NaN            NaN  9.170380e+07   
     GDP                  NaN            NaN  1.935994e+11   

country          Virgin Islands (U.S.)  West Bank and Gaza         World  \
year                                                                       
1960 Population                32000.0                 NaN  3.035056e+09   
     GDP                    24200000.0                 NaN  1.364643e+12   
1961 Population                34100.0                 NaN  3.076121e+09   
  

## We can unstack on different level based on requirements using `level` parameter using Name of actual index

In [216]:
s.unstack('year')
s.unstack('country')

country           Afghanistan       Albania       Algeria       Andorra  \
year                                                                      
1960 Population  8.994793e+06           NaN  1.112489e+07           NaN   
     GDP         5.377778e+08           NaN  2.723638e+09           NaN   
1961 Population  9.164945e+06           NaN  1.140486e+07           NaN   
     GDP         5.488889e+08           NaN  2.434767e+09           NaN   
1962 Population  9.343772e+06           NaN  1.169015e+07           NaN   
...                       ...           ...           ...           ...   
2013 GDP         2.004633e+10  1.278103e+10  2.097035e+11  3.249101e+09   
2014 Population  3.162751e+07  2.893654e+06  3.893433e+07           NaN   
     GDP         2.005019e+10  1.327796e+10  2.135185e+11           NaN   
2015 Population  3.252656e+07  2.889167e+06  3.966652e+07           NaN   
     GDP         1.919944e+10  1.145560e+10  1.668386e+11           NaN   

country                Angola  Antigua and Barbuda    Arab World  \
year                                                               
1960 Population           NaN                  NaN           NaN   
     GDP                  NaN                  NaN           NaN   
1961 Population           NaN                  NaN           NaN   
     GDP                  NaN                  NaN           NaN   
1962 Population           NaN                  NaN           NaN   
...                       ...                  ...           ...   
2013 GDP         1.249121e+11         1.200588e+09  2.846994e+12   
2014 Population  2.422752e+07         9.090000e+04  3.842226e+08   
     GDP         1.267751e+11         1.220976e+09  2.873600e+12   
2015 Population  2.502197e+07         9.181800e+04  3.920223e+08   
     GDP         1.026431e+11         1.297285e+09  2.530102e+12   

country             Argentina       Armenia  Aruba  ...    Uzbekistan  \
year                                                ...                 
1960 Population           NaN           NaN    NaN  ...           NaN   
     GDP                  NaN           NaN    NaN  ...           NaN   
1961 Population           NaN           NaN    NaN  ...           NaN   
     GDP                  NaN           NaN    NaN  ...           NaN   
1962 Population  2.128768e+07           NaN    NaN  ...           NaN   
...                       ...           ...    ...  ...           ...   
2013 GDP         6.239320e+11  1.112147e+10    NaN  ...  5.679566e+10   
2014 Population  4.298003e+07  3.006154e+06    NaN  ...  3.075770e+07   
     GDP         5.480549e+11  1.164444e+10    NaN  ...  6.313285e+10   
2015 Population           NaN  3.017712e+06    NaN  ...  3.129950e+07   
     GDP                  NaN  1.056140e+10    NaN  ...  6.673280e+10   

country               Vanuatu  Venezuela, RB       Vietnam  \
year                                                         
1960 Population           NaN   8.146845e+06           NaN   
     GDP                  NaN   8.607600e+09           NaN   
1961 Population           NaN   8.461684e+06           NaN   
     GDP                  NaN   8.923367e+09           NaN   
1962 Population           NaN   8.790590e+06           NaN   
...                       ...            ...           ...   
2013 GDP         8.017876e+08   3.713366e+11  1.712220e+11   
2014 Population  2.588830e+05            NaN  9.072890e+07   
     GDP         8.149546e+08            NaN  1.862047e+11   
2015 Population           NaN            NaN  9.170380e+07   
     GDP                  NaN            NaN  1.935994e+11   

country          Virgin Islands (U.S.)  West Bank and Gaza         World  \
year                                                                       
1960 Population                32000.0                 NaN  3.035056e+09   
     GDP                    24200000.0                 NaN  1.364643e+12   
1961 Population                34100.0                 NaN  3.076121e+09   
  

------

# 12) The `.unstack()` Method - Part 3

In [219]:
world = pd.read_csv('Data/worldstats.csv', index_col=['country', 'year'])
s = world.stack()
s.head(3)

country     year            
Arab World  2015  Population    3.920223e+08
                  GDP           2.530102e+12
            2014  Population    3.842226e+08
dtype: float64

In [222]:
s.to_frame().head(3)

0
country    year                         
Arab World 2015 Population  3.920223e+08
                GDP         2.530102e+12
           2014 Population  3.842226e+08

## Unstack using list of parameters

In [225]:
s.unstack(level=[1, 0])
s.unstack(level=['year', 'country'])
# we can see year became 1st level and country became 2nd level. Pandas follow order to unstack based on what we provided in the list

year                2015          2014          2013          2012  \
country       Arab World    Arab World    Arab World    Arab World   
Population  3.920223e+08  3.842226e+08  3.765043e+08  3.688026e+08   
GDP         2.530102e+12  2.873600e+12  2.846994e+12  2.773270e+12   

year                2011          2010          2009          2008  \
country       Arab World    Arab World    Arab World    Arab World   
Population  3.610318e+08  3.531122e+08  3.450542e+08  3.368865e+08   
GDP         2.497945e+12  2.103825e+12  1.798878e+12  2.081343e+12   

year                2007          2006  ...          1969          1968  \
country       Arab World    Arab World  ...      Zimbabwe      Zimbabwe   
Population  3.287666e+08  3.209067e+08  ...  5.036321e+06  4.874113e+06   
GDP         1.641666e+12  1.404190e+12  ...  1.747999e+09  1.479600e+09   

year                1967          1966          1965          1964  \
country         Zimbabwe      Zimbabwe      Zimbabwe      Zimbabwe   
Population  4.718612e+06  4.568320e+06  4.422132e+06  4.279561e+06   
GDP         1.397002e+09  1.281750e+09  1.311436e+09  1.217138e+09   

year                1963          1962          1961          1960  
country         Zimbabwe      Zimbabwe      Zimbabwe      Zimbabwe  
Population  4.140804e+06  4.006262e+06  3.876638e+06  3.752390e+06  
GDP         1.159512e+09  1.117602e+09  1.096647e+09  1.052990e+09  

[2 rows x 11211 columns]

In [226]:
s.unstack(level=[0, 1])
s.unstack(level=['country', 'year'])

country       Arab World                                            \
year                2015          2014          2013          2012   
Population  3.920223e+08  3.842226e+08  3.765043e+08  3.688026e+08   
GDP         2.530102e+12  2.873600e+12  2.846994e+12  2.773270e+12   

country                                                             \
year                2011          2010          2009          2008   
Population  3.610318e+08  3.531122e+08  3.450542e+08  3.368865e+08   
GDP         2.497945e+12  2.103825e+12  1.798878e+12  2.081343e+12   

country                                 ...      Zimbabwe                \
year                2007          2006  ...          1969          1968   
Population  3.287666e+08  3.209067e+08  ...  5.036321e+06  4.874113e+06   
GDP         1.641666e+12  1.404190e+12  ...  1.747999e+09  1.479600e+09   

country                                                             \
year                1967          1966          1965          1964   
Population  4.718612e+06  4.568320e+06  4.422132e+06  4.279561e+06   
GDP         1.397002e+09  1.281750e+09  1.311436e+09  1.217138e+09   

country                                                             
year                1963          1962          1961          1960  
Population  4.140804e+06  4.006262e+06  3.876638e+06  3.752390e+06  
GDP         1.159512e+09  1.117602e+09  1.096647e+09  1.052990e+09  

[2 rows x 11211 columns]

### If we look at the following data, we can see for some country(such as Albania) there are no data for 1960, 1961,etc.
This is because for those year there is no data(GDP and Population) for that country. So when we unstack year, those empty values became NaN.

In [227]:
s.unstack('year')

year                            1960          1961          1962  \
country                                                            
Afghanistan Population  8.994793e+06  9.164945e+06  9.343772e+06   
            GDP         5.377778e+08  5.488889e+08  5.466667e+08   
Albania     Population           NaN           NaN           NaN   
            GDP                  NaN           NaN           NaN   
Algeria     Population  1.112489e+07  1.140486e+07  1.169015e+07   
...                              ...           ...           ...   
Yemen, Rep. GDP                  NaN           NaN           NaN   
Zambia      Population  3.049586e+06  3.142848e+06  3.240664e+06   
            GDP         6.987397e+08  6.823597e+08  6.792797e+08   
Zimbabwe    Population  3.752390e+06  3.876638e+06  4.006262e+06   
            GDP         1.052990e+09  1.096647e+09  1.117602e+09   

year                            1963          1964          1965  \
country                                                            
Afghanistan Population  9.531555e+06  9.728645e+06  9.935358e+06   
            GDP         7.511112e+08  8.000000e+08  1.006667e+09   
Albania     Population           NaN           NaN           NaN   
            GDP                  NaN           NaN           NaN   
Algeria     Population  1.198513e+07  1.229597e+07  1.262695e+07   
...                              ...           ...           ...   
Yemen, Rep. GDP                  NaN           NaN           NaN   
Zambia      Population  3.342894e+06  3.449266e+06  3.559687e+06   
            GDP         7.043397e+08  8.226397e+08  1.061200e+09   
Zimbabwe    Population  4.140804e+06  4.279561e+06  4.422132e+06   
            GDP         1.159512e+09  1.217138e+09  1.311436e+09   

year                            1966          1967          1968  \
country                                                            
Afghanistan Population  1.014884e+07  1.036860e+07  1.059979e+07   
            GDP         1.400000e+09  1.673333e+09  1.373333e+09   
Albania     Population           NaN           NaN           NaN   
            GDP                  NaN           NaN           NaN   
Algeria     Population  1.298027e+07  1.335420e+07  1.374438e+07   
...                              ...           ...           ...   
Yemen, Rep. GDP                  NaN           NaN           NaN   
Zambia      Population  3.674088e+06  3.792864e+06  3.916928e+06   
            GDP         1.239000e+09  1.340639e+09  1.573739e+09   
Zimbabwe    Population  4.568320e+06  4.718612e+06  4.874113e+06   
            GDP         1.281750e+09  1.397002e+09  1.479600e+09   

year                            1969  ...          2006          2007  \
country                               ...                               
Afghanistan Population  1.084951e+07  ...  2.518362e+07  2.587754e+07   
            GDP         1.408889e+09  ...  7.057598e+09  9.843842e+09   
Albania     Population           NaN  ...  2.992547e+06  2.970017e+06   
            GDP                  NaN  ...  8.992642e+09  1.070101e+10   
Algeria     Population  1.414444e+07  ...  3.374933e+07  3.426197e+07   
...                              ...  ...           ...           ...   
Yemen, Rep. GDP                  NaN  ...  1.908173e+10  2.563367e+10   
Zambia      Population  4.047479e+06  ...  1.238151e+07  1.273868e+07   
            GDP         1.926399e+09  ...  1.275686e+10  1.405696e+10   
Zimbabwe    Population  5.036321e+06  ...  1.312794e+07  1.329780e+07   
            GDP         1.747999e+09  ...  5.443896e+09  5.291950e+09   

year                            2008          2009          2010  \
country                                                            
Afghanistan Population  2.652874e+07  2.720729e+07  2.796221e+07   
            GDP         1.019053e+10  1.248694e+10  1.593680e+10   
Albania     Population  2.947314e+06  2.927519e+06  2.913021e+06   
            GDP         1.288135e+10  1.204421e+10  1.1

## We can use `fill_value` to fill in those empty values.

In [229]:
s = s.unstack('year', fill_value=0)
s.head(3)

year                            1960          1961          1962  \
country                                                            
Afghanistan Population  8.994793e+06  9.164945e+06  9.343772e+06   
            GDP         5.377778e+08  5.488889e+08  5.466667e+08   
Albania     Population  0.000000e+00  0.000000e+00  0.000000e+00   

year                            1963          1964          1965  \
country                                                            
Afghanistan Population  9.531555e+06  9.728645e+06  9.935358e+06   
            GDP         7.511112e+08  8.000000e+08  1.006667e+09   
Albania     Population  0.000000e+00  0.000000e+00  0.000000e+00   

year                            1966          1967          1968  \
country                                                            
Afghanistan Population  1.014884e+07  1.036860e+07  1.059979e+07   
            GDP         1.400000e+09  1.673333e+09  1.373333e+09   
Albania     Population  0.000000e+00  0.000000e+00  0.000000e+00   

year                            1969  ...          2006          2007  \
country                               ...                               
Afghanistan Population  1.084951e+07  ...  2.518362e+07  2.587754e+07   
            GDP         1.408889e+09  ...  7.057598e+09  9.843842e+09   
Albania     Population  0.000000e+00  ...  2.992547e+06  2.970017e+06   

year                            2008          2009          2010  \
country                                                            
Afghanistan Population  2.652874e+07  2.720729e+07  2.796221e+07   
            GDP         1.019053e+10  1.248694e+10  1.593680e+10   
Albania     Population  2.947314e+06  2.927519e+06  2.913021e+06   

year                            2011          2012          2013  \
country                                                            
Afghanistan Population  2.880917e+07  2.972680e+07  3.068250e+07   
            GDP         1.793024e+10  2.053654e+10  2.004633e+10   
Albania     Population  2.904780e+06  2.900247e+06  2.896652e+06   

year                            2014          2015  
country                                             
Afghanistan Population  3.162751e+07  3.252656e+07  
            GDP         2.005019e+10  1.919944e+10  
Albania     Population  2.893654e+06  2.889167e+06  

[3 rows x 56 columns]

-------

# 13) The `.pivot()` Method
+ **re-orient the data set**
+ basically take values currently in column to column headers

In [239]:
sales = pd.read_csv('Data/salesmen.csv', parse_dates=['Date'])
sales['Salesman'] = sales['Salesman'].astype('category')
sales.head(3)

Date Salesman  Revenue
0 2016-01-01      Bob     7172
1 2016-01-02      Bob     6362
2 2016-01-03      Bob     5982

In [235]:
len(sales)

1830

In [236]:
sales['Salesman'].value_counts()

Ronald    366
Dave      366
Jeb       366
Bob       366
Oscar     366
Name: Salesman, dtype: int64

As for our case, we have only 5 Salesman. Instead of storing data in current format which gonna take a lot of memory usage, we can re-orient the dataset.

### we can condense the data as follow
+ we can see original dataframe of 1830 is reduced down to 366.

In [242]:
len(sales.pivot(index='Date', columns='Salesman', values='Revenue'))

366

In [243]:
sales.pivot(index='Date', columns='Salesman', values='Revenue')

Salesman     Bob  Dave   Jeb  Oscar  Ronald
Date                                       
2016-01-01  7172  1864  4430   5250    2639
2016-01-02  6362  8278  8026   8661    4951
2016-01-03  5982  4226  5188   7075    2703
2016-01-04  7917  3868  3144   2524    4258
2016-01-05  7837  2287   938   2793    7771
...          ...   ...   ...    ...     ...
2016-12-27  2045  2843  6666    835    2981
2016-12-28   100  8888  1243   3073    6129
2016-12-29  4115  9490  3498   6424    7662
2016-12-30  2577  3594  8858   7088    2570
2016-12-31  3845  6830  9717   8408    2619

[366 rows x 5 columns]

--------

# 14) The `.pivot_table()` Method

In [245]:
foods = pd.read_csv('Data/foods.csv')
foods.head(3)

First Name  Gender      City Frequency     Item  Spend
0      Wanda  Female  Stamford    Weekly   Burger  15.66
1       Eric    Male  Stamford     Daily  Chalupa  10.56
2    Charles    Male  New York     Never    Sushi  42.14

### What are the average spending and Total Spending per Gender?

In [247]:
foods.pivot_table(index='Gender', values='Spend', aggfunc='mean')

Spend
Gender           
Female  50.709629
Male    49.397623

In [248]:
foods.pivot_table(index='Gender', values='Spend', aggfunc='sum')

Spend
Gender          
Female  25963.33
Male    24106.04

### What about other sales?

In [249]:
foods.pivot_table(index='Item', values='Spend', aggfunc='sum')

Spend
Item              
Burger     7765.73
Burrito    8270.44
Chalupa    7644.52
Donut      8758.76
Ice Cream  8886.99
Sushi      8742.93

### We can also make multi Index 

In [250]:
foods.pivot_table(index=['Gender', 'Item'], values='Spend', aggfunc='sum')

Spend
Gender Item              
Female Burger     4094.30
       Burrito    4257.82
       Chalupa    4152.26
       Donut      4743.00
       Ice Cream  4032.87
       Sushi      4683.08
Male   Burger     3671.43
       Burrito    4012.62
       Chalupa    3492.26
       Donut      4015.76
       Ice Cream  4854.12
       Sushi      4059.85

In [251]:
foods.pivot_table(index=['Gender', 'Item'], columns=['City'], values='Spend', aggfunc='sum')

City              New York  Philadelphia  Stamford
Gender Item                                       
Female Burger      1239.04       1639.24   1216.02
       Burrito      978.95       1458.76   1820.11
       Chalupa      876.58       1673.33   1602.35
       Donut       1446.78       1639.26   1656.96
       Ice Cream   1521.62       1479.22   1032.03
       Sushi       1480.29       1742.88   1459.91
Male   Burger      1294.09        938.18   1439.16
       Burrito     1399.40       1312.93   1300.29
       Chalupa     1227.77       1114.23   1150.26
       Donut       1345.27       1249.36   1421.13
       Ice Cream   1603.63       2191.27   1059.22
       Sushi       1396.15       1395.88   1267.82

In [252]:
foods.pivot_table(index=['Gender', 'Item'], columns=['Frequency', 'City'], values='Spend', aggfunc='sum')

Frequency           Daily                        Monthly               \
City             New York Philadelphia Stamford New York Philadelphia   
Gender Item                                                             
Female Burger      262.67       231.68   144.66   171.86       215.05   
       Burrito     224.45       321.57   195.63   122.74        34.28   
       Chalupa      43.19        23.49    95.70   158.37       289.96   
       Donut       478.10       247.40   124.35   284.53        50.25   
       Ice Cream   262.19       177.69    92.88    92.53        74.51   
       Sushi        81.07       348.53   393.27    93.16        78.71   
Male   Burger      319.46       112.70   197.72   187.29       213.14   
       Burrito     236.21       165.76   276.23   147.54        89.58   
       Chalupa      54.09       274.81   192.64   333.76        90.70   
       Donut       230.00       190.71   129.42   183.73       259.29   
       Ice Cream   178.22       148.69    15.17   152.20       386.55   
       Sushi       225.57       280.93      NaN   430.86       155.59   

Frequency                    Never                          Often  ...  \
City             Stamford New York Philadelphia Stamford New York  ...   
Gender Item                                                        ...   
Female Burger      238.49    97.89       218.97    90.97    94.96  ...   
       Burrito      67.94   189.73       382.30   366.34   103.60  ...   
       Chalupa     161.98    35.15       121.97   156.36    39.73  ...   
       Donut       229.30    56.07       433.58   157.33   130.63  ...   
       Ice Cream   125.85   206.15       156.03    77.66   232.26  ...   
       Sushi       108.39    69.33       190.58    19.56   343.94  ...   
Male   Burger       27.16    90.32        17.31      NaN    55.47  ...   
       Burrito     119.60    86.78       378.32   351.84   189.92  ...   
       Chalupa     343.76   199.09       242.98   138.70   251.52  ...   
       Donut       119.53   219.63        53.65   219.64   139.80  ...   
       Ice Cream    74.36    57.54       321.94   119.77   347.55  ...   
       Sushi        93.03   164.01       245.77   204.90     9.31  ...   

Frequency            Once   Seldom                         Weekly  \
City             Stamford New York Philadelphia Stamford New York   
Gender Item                                                         
Female Burger       95.05    63.16       409.05    97.53   184.35   
       Burrito     168.01   167.54       198.11   234.49    39.69   
       Chalupa      40.59   160.00       274.51   175.25   171.52   
       Donut       158.24    90.81       183.25   104.66   285.56   
       Ice Cream   279.33   242.35       203.10   117.73   227.62   
       Sushi       194.75    87.70       314.24    27.82   410.89   
Male   Burger      326.64   225.68       188.06   159.75    69.69   
       Burrito      30.15   202.40        27.71     9.84   256.74   
       Chalupa        NaN    11.69       130.75   174.02   163.20   
       Donut       139.89    16.25        99.01   163.31    74.44   
       Ice Cream   182.71   270.42       138.44    40.83   223.86   
       Sushi       314.65   160.49        91.26   224.15   150.15   

Frequency                                Yearly                        
City             Philadelphia Stamford New York Philadelphia Stamford  
Gender Item                                                            
Female Burger           32.00   155.02   259.30       123.17   307.03  
       Burrito          62.82   230.91    71.26       116.75   172.98  
       Chalupa          84.41   204.69   157.82       504.44   348.16  
       Donut           208.80   440.06    62.95       175.23   224.48  
       Ice Cream       285.28   125.03   151.67       159.86    15.24  
       Sushi           334.00   315.36   232.41       145.85   362.74  
Male   Burger           99.89   310.21    49.61       197.36   225.07  
       Burrito         289.25   161.8

## Other avaliable `aggfunc`

In [256]:
foods.pivot_table(index=['Gender', 'Item'], columns='City', values='Spend', aggfunc='count')
foods.pivot_table(index=['Gender', 'Item'], columns='City', values='Spend', aggfunc='max')
foods.pivot_table(index=['Gender', 'Item'], columns='City', values='Spend', aggfunc='min')

City              New York  Philadelphia  Stamford
Gender Item                                       
Female Burger         2.25          1.97      6.24
       Burrito        1.02          1.04      1.18
       Chalupa        1.96          9.35      9.09
       Donut          3.15          2.13      1.68
       Ice Cream     13.39          7.61      8.80
       Sushi          2.52         11.68      8.20
Male   Burger         5.43          1.71      2.83
       Burrito       15.90          8.58      3.64
       Chalupa       11.61          1.94     10.56
       Donut          1.49          1.26      6.63
       Ice Cream     14.06          4.89      3.43
       Sushi          3.28          2.01     32.15

## Directly performing with pandas, insteading of directly calling on DataFrame

In [258]:
pd.pivot_table(data=foods, index=['Gender', 'Item'], columns='City', values='Spend', aggfunc='count')

City              New York  Philadelphia  Stamford
Gender Item                                       
Female Burger           24            31        27
       Burrito          23            28        34
       Chalupa          19            32        25
       Donut            31            30        34
       Ice Cream        27            32        22
       Sushi            31            30        32
Male   Burger           22            21        31
       Burrito          25            30        28
       Chalupa          25            23        23
       Donut            30            33        29
       Ice Cream        29            41        25
       Sushi            27            28        18

-----

# 15) The `pd.melt()` Method
**take the aggregated data set into Tabular format.**

+ `id_vars`: columns which are going to be preserved.

In [260]:
sales = pd.read_csv('Data/quarters.csv')
sales

Salesman      Q1      Q2      Q3      Q4
0    Boris  602908  233879  354479   32704
1      Bob   43790  514863  297151  544493
2    Tommy  392668  113579  430882  247231
3   Travis  834663  266785  749238  570524
4   Donald  580935  411379  110390  651572
5      Ted  656644   70803  375948  321388
6      Jeb  486141  600753  742716  404995
7    Stacy  479662  742806  770712    2501
8   Morgan  992673  879183   37945  293710

In [264]:
pd.melt(sales, id_vars='Salesman', var_name='Quarter', value_name='Revenue') # we want to preserved Salesman

Salesman Quarter  Revenue
0     Boris      Q1   602908
1       Bob      Q1    43790
2     Tommy      Q1   392668
3    Travis      Q1   834663
4    Donald      Q1   580935
5       Ted      Q1   656644
6       Jeb      Q1   486141
7     Stacy      Q1   479662
8    Morgan      Q1   992673
9     Boris      Q2   233879
10      Bob      Q2   514863
11    Tommy      Q2   113579
12   Travis      Q2   266785
13   Donald      Q2   411379
14      Ted      Q2    70803
15      Jeb      Q2   600753
16    Stacy      Q2   742806
17   Morgan      Q2   879183
18    Boris      Q3   354479
19      Bob      Q3   297151
20    Tommy      Q3   430882
21   Travis      Q3   749238
22   Donald      Q3   110390
23      Ted      Q3   375948
24      Jeb      Q3   742716
25    Stacy      Q3   770712
26   Morgan      Q3    37945
27    Boris      Q4    32704
28      Bob      Q4   544493
29    Tommy      Q4   247231
30   Travis      Q4   570524
31   Donald      Q4   651572
32      Ted      Q4   321388
33      Jeb      Q4   404995
34    Stacy      Q4     2501
35   Morgan      Q4   293710

--------